File: Tracker/src/main/java/com/cyrillrx/tracker/consumer/EventConsumer.java
package com.cyrillrx.tracker.consumer;

import com.cyrillrx.tracker.TrackerChild;
import com.cyrillrx.tracker.event.TrackEvent;

import java.util.Collection;

/**
 * @author Cyril Leroux
 *         Created on 20/04/16.
 */
public abstract class EventConsumer<EventContainer extends Collection<TrackEvent>>
        implements Runnable {

    protected static final TrackEvent STOP_EVENT = new TrackEvent.Builder()
            .setCategory("STOP_EVENT")
            .build();

    protected final TrackerChild tracker;
    protected final EventContainer events;
    protected boolean running;

    public EventConsumer(TrackerChild tracker, EventContainer events) {
        this.tracker = tracker;
        this.events = events;
    }

    protected abstract void consume();
}

File: tracker/src/main/java/com/cyrillrx/tracker/consumer/EventConsumer.kt
package com.cyrillrx.tracker.consumer

import com.cyrillrx.tracker.event.TrackEvent
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.channels.ReceiveChannel
import kotlinx.coroutines.launch
import java.util.Queue

/**
 * @author Cyril Leroux
 *         Created on 20/04/2016.
 */
abstract class EventConsumer(protected val events: Channel<TrackEvent>) {

    val coroutineScope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    var currentJob: Job? = null

    protected var running = false

    fun start(): Job = currentJob ?: startNewJob().also { currentJob = it }

    suspend fun stop(): Boolean {

        if (!running) return false
        val job = currentJob ?: return false
        if (job.isCancelled || job.isCompleted) return false

        events.send(STOP_EVENT)

        return true
    }

    private fun startNewJob(): Job = coroutineScope.launch {
        running = true
        while (running) {
            consume()
        }
    }

    protected abstract suspend fun consume()

    companion object {

        internal val STOP_EVENT = TrackEvent.Builder()
            .setCategory("STOP_EVENT")
            .setName("STOP_EVENT")
            .build()
    }
}